### ISSUE TITLE
Bad inference using overloaded `__or__` with generic value and type

### ISSUE THREAD

**Issue by RocketRace** [microsoft/pyright#3761]
**Describe the bug**
When one argument of `|` is a generic value, and the other argument is a type (being used as e.g. a callable), if the types are not  immediately compatible with the signature of `__or__`, the result is inferred as `UnionType`.

**To Reproduce**
```py
from __future__ import annotations
from typing import Callable, Generic, TypeVar

A = TypeVar("A")
B = TypeVar("B")
C = TypeVar("C")

class T(Generic[A, B]):
    def __or__(self, f: Callable[[B], C]) -> T[A, C]: ...

def foo0(t: T[A, B]) -> T[A, str]:
    return reveal_type(t | str.upper) # Error, acceptable (str cannot be assigned to B)

def foo1(t: T[A, B]) -> T[A, str]:
    return reveal_type(t | str) # Pass, as expected (object can be assigned to B)

def foo2(t: T[A, B]) -> T[A, int]:
    return reveal_type(t | int) # Type incorrectly inferred as UnionType
```

**Expected behavior**

A type error (because `str | SupportsInt | ...` cannot be assigned to `B`) 

(Or possibly stronger bidirectional inference? See Further Discussion)

**VS Code extension or command-line**

* Pylance v2022.7.40
* pyright 1.1.264 (via `npx`)

**Further discussion**

As an extension, would it also be "correct" behavior to narrow `A` and `B` such that the types are compatible? I believe the example would type check in some other languages like Rust or Haskell, but I haven't checked. Maybe they have stronger bidirectional inference. Either way, the result shouldn't be inferred as `UnionType`.

**Comment by erictraut** [comment]
I think pyright is doing the right thing here.

The expression `t | int` is the equivalent of `t.__or__(int)` or `int.__ror__(t)`. Since the expression `t.__or__(int)` results in a type violation, pyright falls back on `int.__ror__(t)`. This evaluates without a type violation to `UnionType` by virtual of the `__ror__` method defined on the `type` class.

```python
class type:
    def __ror__(self, __t: Any) -> types.UnionType: ...
```

**Comment by RocketRace** [comment]
That's unfortunate, since it can lead some some hard to debug results (I personally was making a function composition pipeline). I guess a typeshed change isn't possible either (`__ror__` currently accepts `Any`)?
